src/hivpy/hiv_testing.py

```diff
@@ -49,163 +55,243 @@ def __init__(self, **kwargs):
-                                             self.calc_testing_outcomes,
-                                             sub_pop=prev_tested_population)
-                self.apply_test_outcomes_to_sub_pop(pop, tested, prev_tested_population)
+        return marked
 
-    def apply_test_outcomes_to_sub_pop(self, pop, tested, sub_pop):
+    def calc_symptomatic_prob_test(self, adc_tm1, tb_tm1, tb_tm2, non_tb_who3_tm1):
         """
-        Uses HIV testing outcomes for a given sub-population to
-        set last test date and reset number of partners since last test.
+        Calculates the probability of being tested for a group
+        with specific symptoms and returns it. Presence of
+        an AIDS defining condition (ADC; any WHO4) in the previous time step,
+        tuberculosis (TB) in the previous two time steps,
+        and a non-TB WHO3 disease in the previous time step
+        all affect groupings and test probability.
         """
-        # set last test date
-        pop.set_present_variable(col.LAST_TEST_DATE, pop.date,
-                                 sub_pop=pop.apply_bool_mask(tested, sub_pop))
-        # "reset" dummy partner columns
-        pop.set_present_variable(col.NSTP_LAST_TEST, 0,
-                                 sub_pop=pop.apply_bool_mask(tested, sub_pop))
-        pop.set_present_variable(col.NP_LAST_TEST, 0,
-                                 sub_pop=pop.apply_bool_mask(tested, sub_pop))
+        # assume asymptomatic by default
+        prob_test = 0
+        # presence of ADC last time step
+        if adc_tm1:
+            prob_test = self.prob_test_who4
+        # presence of TB last time step but not the time step before, no ADC last time step
+        elif tb_tm1 and not tb_tm2:
+            prob_test = self.prob_test_tb
+        # presence of a non-TB WHO3 disease last time step, no ADC or TB last time step
+        elif non_tb_who3_tm1 and not tb_tm1:
+            prob_test = self.prob_test_non_tb_who3
 
-    def update_sub_pop_test_date(self, pop, sub_pop, prob_test):
+        return prob_test
+
+    def test_mark_non_hiv_symptomatic(self, pop):
         """
-        Update the last test date of a sub-population based on a given probability.
+        Mark non-HIV symptomatic individuals to undergo testing this time step.
         """
-        if len(sub_pop) > 0:
-            r = rng.uniform(size=len(sub_pop))
-            tested = r < prob_test
-            pop.set_present_variable(col.LAST_TEST_DATE, pop.date,
-                                     sub_pop=pop.apply_bool_mask(tested, sub_pop))
+        # testing occurs after a certain year if there is no covid disruption
+        if ((pop.date.year >= self.date_start_testing)
+           & (not (self.covid_disrup_affected | self.testing_disrup_covid))):
+
+            # undiagnosed (last time step) population not scheduled for testing
+            not_diag_tested_pop = pop.get_sub_pop([(pop.get_correct_column(col.HIV_DIAGNOSED, dt=1), op.eq, False),
```

Comment:
No particular reason to use diagnosis of HIV from the last time step here, diagnosis has not yet been updated and it would be helpful to not need to track multiple time steps worth of it. If HIV diagnosis date is relevant, there is a separate column for this information.